dpkg/distroless_scanner.go

```diff
@@ -0,0 +1,145 @@
+
+// Kind implements scanner.VersionedScanner.
+func (ps *DistrolessScanner) Kind() string { return distrolessKind }
+
+// Scan attempts to find a dpkg database files in the layer and read all
+// of the installed packages it can find. These files are found in the
+// dpkg/status.d directory.
+//
+// It's expected to return (nil, nil) if there's no dpkg databases in the layer.
+//
+// It does not respect any dpkg configuration files.
+func (ps *DistrolessScanner) Scan(ctx context.Context, layer *claircore.Layer) ([]*claircore.Package, error) {
+	defer trace.StartRegion(ctx, "Scanner.Scan").End()
+	trace.Log(ctx, "layer", layer.Hash.String())
+	ctx = zlog.ContextWithValues(ctx,
+		"component", "dpkg/DistrolessScanner.Scan",
+		"version", ps.Version(),
+		"layer", layer.Hash.String())
+	zlog.Debug(ctx).Msg("start")
+	defer zlog.Debug(ctx).Msg("done")
+
+	rd, err := layer.Reader()
+	if err != nil {
+		return nil, fmt.Errorf("opening layer failed: %w", err)
+	}
+	defer rd.Close()
+	sys, err := tarfs.New(rd)
+	if err != nil {
+		return nil, fmt.Errorf("opening layer failed: %w", err)
+	}
+
+	var pkgs []*claircore.Package
+	walk := func(p string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.Name() == "status.d" && d.IsDir() {
+			zlog.Debug(ctx).Str("path", p).Msg("found potential distroless dpkg db directory")
+			dbFiles, err := sys.ReadDir(p)
+			if err != nil {
+				return fmt.Errorf("error reading DB directory: %w", err)
+			}
+			for _, f := range dbFiles {
+				pkgCt := 0
+				fn := filepath.Join(p, f.Name())
+				ctx = zlog.ContextWithValues(ctx, "database-file", fn)
+				zlog.Debug(ctx).Msg("examining package database")
+				db, err := sys.Open(fn)
+				if err != nil {
+					return fmt.Errorf("reading database files from layer failed: %w", err)
+				}
+
+				// The database is actually an RFC822-like message with "\n\n"
+				// separators, so don't be alarmed by the usage of the "net/textproto"
+				// package here.
+				tp := textproto.NewReader(bufio.NewReader(db))
+			Restart:
+				hdr, err := tp.ReadMIMEHeader()
+				for ; (err == nil || errors.Is(err, io.EOF)) && len(hdr) > 0; hdr, err = tp.ReadMIMEHeader() {
+					name := hdr.Get("Package")
+					v := hdr.Get("Version")
+					p := &claircore.Package{
+						Name:      name,
+						Version:   v,
+						Kind:      claircore.BINARY,
+						Arch:      hdr.Get("Architecture"),
+						PackageDB: fn,
+					}
+					if src := hdr.Get("Source"); src != "" {
+						p.Source = &claircore.Package{
+							Name: src,
+							Kind: claircore.SOURCE,
+							// Right now, this is an assumption that discovered source
+							// packages relate to their binary versions. We see this in
+							// Debian.
+							Version:   v,
+							PackageDB: fn,
+						}
+					}
+					pkgCt++
+					pkgs = append(pkgs, p)
+				}
+				switch {
+				case errors.Is(err, io.EOF):
+				default:
+					zlog.Warn(ctx).Err(err).Msg("unable to read entry")
+					goto Restart
```

Comment:
can this potentially infinitely loop?

Reply:
I don't believe so, we're going to out when we get to the EOF, you might be seeing something I'm not

Reply:
I just wonder if it's possible the underlying reader returns non-EOF errors without reading anything, so we keep trying over and over again due to some underlying filesystem issues or something. Maybe I'm overthinking it?

Reply:
Something like this: https://go.dev/play/p/2NKKCBv3PUp

Reply:
I guess that is possible, could do something like this:
```go
switch {
case errors.Is(err, io.EOF):
default:
    if _, ok := err.(textproto.ProtocolError); ok {
        goto Restart
    }
    zlog.Warn(ctx).Err(err).Msg("unable to read entry")
}
```
so we only carry on